### ISSUE TITLE
HTTP timeouts no longer work.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8481]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8481|
|Type|release blocker: regression|
|Created|2016-06-16 14:10:41Z|
|Branch|https://github.com/twisted/twisted/tree/8481-lukasa-issue|

This bug was introduced by the abstraction layer for HTTP/1.1 and HTTP/2 protocols, the `_GenericHTTPChannelProtocol`.

The `HTTPChannelFactory` currently sets `protocol.timeOut`. That used to set `timeOut` on `HTTPChannel`, but since we moved to `_GenericHTTPChannelProtocol` it now sets `timeOut` on `_GenericHTTPChannelProtocol`, which *does not* pass this through to its underlying `HTTPChannel` (or progress it onto the `H2Connection`).

This means that we...kinda broke timeouts. We should maybe fix that.

<details><summary>Searchable metadata</summary>

```
trac-id__8481 8481
type__release_blocker__regression release blocker: regression
reporter__Lukasa Lukasa
priority__normal normal
milestone__Twisted_16_3 Twisted 16.3
branch__8481_lukasa_issue 8481-lukasa-issue
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1466086241002036 1466086241002036
changetime__1467117066846462 1467117066846462
version__None None
owner__hawkowl hawkowl
cc__hawkowl
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

whoops

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_3"></a>|@Lukasa commented|
|-|-|

A fix for this is available in pull request 193: https://github.com/twisted/twisted/pull/193

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

I don't undertand why we need to keep a copy of the timeout in the protocol, and why not just proxy it to the channel.

Maybe we can add the timeOut attribute to a new IProtocolWithTimeout interface and the protocol will just proxy this interface.

@lukasa, I know that you answer to my concern over IRC but I think that we need to have the main discussion here.

I am leaving this in the review queue as someone else can review it.

I am -1 for this change as I think that timeOut should be just proxied, without having a copy which might get out of sync :)

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_5"></a>|@Lukasa commented|
|-|-|

> I don't undertand why we need to keep a copy of the timeout in the protocol, and why not just proxy it to the channel.

We use a copy of several different properties in this class so that we can create a new `H2Connection` and put them in the right place. These things are all set on the `HTTPChannel` by either `HTTPFactory` or its subclasses (mostly `Site`), and need to be propagated to the `H2Connection` so that `Request` objects can get access to them.

We don't have to do it this way: for example, we could just proxy through to the `HTTPChannel` and then copy the properties off the `HTTPChannel` to the `H2Channel` when needed, but that would require defining what amounts to a single-use interface that has no sensible cohesion to it: it's a grab bag of things that `HTTPFactory` decided it needed to be able to set on the protocols it created.

I have no aversion to doing that per se, but it amounts to rewriting this class entirely rather than incrementally fixing. That's up to the Twisted maintainers.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

Sorry for the trouble.

I now think that your patch is OK.

@lukasa, since you have write access to Twisted repo, for future patches please create a branch in the main repo and send it to buildbot.

I am now waiting for the build results.

--------

I don't like the way the H2Connection is initialized, but I think this should be solved in a separate ticket.


Instead of

```
transport = self._channel.transport
self._channel = H2Connection()
self._channel.requestFactory = self._requestFactory
self._channel.site = self._site
self._channel.factory = self._factory
self._channel.makeConnection(transport)
```

Why not have

```
rootChannel = self._channel
self._channel = H2Connection(
    requestFactory=rootChannel.requestFactory,
    timeOut=rootChannel.timeOut,
    # More here
    )
self._channel.makeConnection(rootChannel.transport)
```

or

```
rootChannel = self._channel
self._channel = H2Connection(rootChannel)
self._channel.makeConnection(rootChannel.transport)
```


And the in `H2Connection.__init__` we can take care of setting up the right state and make sure that we set all initial variables

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

buildbot is green.

Regarding this comment.

```
# Maybe we should fix this?
```

I think that this should be addressed in a ticket, rather than left as a comment in a test.

... getting late here so in case it is not already reviewed I will continue later

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa commented|
|-|-|

So the initialization of H2Connection is definitely tricky. Again, the reason that this is done is for compat with the current method of initializing HTTPChannel: that is, it's incremental, and some of these properties are patched onto it. The goal here is to have H2Connection be as close as possible in API terms to HTTPChannel.

Agree that "maybe we should fix this" should be a ticket, has been opened as [#8488](https://github.com/twisted/twisted/issues/8488). I've also removed the "maybe we should fix this" comment in favour of that ticket.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl set owner to @hawkowl|
|-|-|

LGTM.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset bb0d1d67198c3ae8ce8bb1f32816eef4c0a98dc6
```
#!CommitTicketReference repository="" revision="bb0d1d67198c3ae8ce8bb1f32816eef4c0a98dc6"
Merge 8481-lukasa-issue: Fix HTTP timeouts.

Author: lukasa
Reviewers: adiroiban, hawkowl
Fixes: #8481
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

In changeset 23751e676545a60825443d6bd377748c17145dab
```
#!CommitTicketReference repository="" revision="23751e676545a60825443d6bd377748c17145dab"
Merge 8481-lukasa-issue: Fix HTTP timeouts.

Author: lukasa
Reviewers: adiroiban, hawkowl
Fixes: #8481
```
